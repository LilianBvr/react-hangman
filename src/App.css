.hangman{

	margin: 3vh 0 3vh 0; padding: 0;
	height: 94vh;

	display: flex;
	flex-wrap: nowrap;
	flex-direction: column;
	justify-content:space-around;
	align-items: center;

	font-family: sans-serif;
	background-color: #DDDDDDDD;

	user-select: none;
}

.hangman .restart{
	margin:5px;
	min-width: 40px;min-height: 40px;

	background-color: #DDDDDDDD;
  	border: 2px solid #888;
  	border-radius: 4px;

  	font-family:arial;
  	font-size:1.2em;
  	color: #888;
  	text-align: center;
}

.hangman > .restart:hover{
	background-color:  #888;
	color:white;
	transition: all 0.3s;
	cursor: pointer;
}
