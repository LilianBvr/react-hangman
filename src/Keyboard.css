.keyboard{
	margin:0;
	padding:0;
	max-width:55vw;

	display:flex;
	flex-wrap: wrap;
	justify-content: center;
}

.keyboard > .letter{
	width:4vw; height: 4vw;
	margin:5px;
	min-width: 40px;min-height: 40px;

	background-color: #DDDDDDDD;
  	border: 2px solid #888;
  	border-radius: 4px;

  	font-family:arial;
  	color: #888;
  	text-align: center;
}

.keyboard > .letter:hover{
	background-color:  #888;
	color:white;
	transition: all 0.3s;
	cursor: pointer;
}

.keyboard > .letter.tested{
	background-color:  #888;
	color:white;
}

.keyboard > .letter.tested:hover{
	cursor:not-allowed;
}
