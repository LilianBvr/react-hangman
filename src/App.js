import React, {Component} from 'react';

import './App.css';

import Mask from './Mask'
import Keyboard from './Keyboard'

export const LETTERS = Array.from('ABCDEFGHIJKLMNOPQRSTUVWXYZ')

class App extends Component {

	constructor(){
		super()

		/*	Etat local	*/
		this.state = this.getInitState()
	}
	
	/*	Méthode de rendu du composant	*/
	render(){
		const { mask, testedLetters, won } = this.state
	  	return (
	    	<div className="hangman">
	    		<Mask mask={mask} />

	    		{
	    			won ? <button className="restart" onClick={this.handleRestartClick}>
							Recommencer
						</button> :
	    				<Keyboard 
	    					letters={LETTERS} 
	    					testedLetters={testedLetters}
	    					onClick={this.handleLetterClick}
	    				/>
	    		}

	    	</div>
	  	)
	}

	/*	Gestionnaire de clic sur une lettre 	*/
	/* @autobind */
	handleLetterClick = letter => {
		const { mask, testedLetters, word } = this.state
		const newMask = this.revealLetter(letter)

		if(!testedLetters.includes(letter))
		{
			testedLetters.push(letter)

			this.setState({
				testedLetters: testedLetters,
				mask: newMask,
				won: word.every((char, index) => char === newMask[index] )
			})
		}
	}

	/*	Gestionnaire de l'évènement Recommencer*/
	/* @autobind */
	handleRestartClick = () => { 
		this.setState( this.getInitState() ) 
	}

	/*	Révèle la lettre en paramètres et retourne le nouveau masque	*/
	revealLetter( letter ){
		const { mask, word } = this.state

		return mask.map((char, i) => letter===word[i] ? word[i] : char)
	}

	/*	Renvoie un mot aléatoire de la liste locale	*/
	getRandomWord(){
		return Array.from(WORDS[ Math.floor(Math.random() * WORDS.length )]);
	}

	/*	Renvoie l'état initial	*/
	getInitState(){
		const word = this.getRandomWord()

		return { 
			word: word,
			mask: this.getInitialMask(word),
			testedLetters: [],
			won: false
		}
	}

	/*	Méthode retournant le mot masqué de départ	*/
	getInitialMask(word){
		return word.map((letter, index)=>[0,word.length-1].includes(index) ? letter : '_')
	}
}


const WORDS = [
	'MANGER',
	'BOIRE',
	'DORMIR',
	'POMME',
	'POIRE',
	'FRAISE',
	'FRAMBOISE',
	'SUPREME',
	'PALACE',
	'GUCCI'
]


export default App;
