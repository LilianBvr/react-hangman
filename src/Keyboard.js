import React from 'react'
import PropTypes from 'prop-types'

import './Keyboard.css'

const Keyboard = ({ letters, testedLetters, onClick}) => (

	<div className='keyboard'>
		{
			letters.map(( letter, index ) => (
				<button 
					className={`letter ${testedLetters.includes(letter) ? 'tested' : ''}`}
					key={index}
					onClick={() => onClick(letter)}
				>
					{letter}
				</button>
			))
		}
	</div>
)

Keyboard.propTypes = {
	letters: PropTypes.array,
	testedLetters: PropTypes.array
}

export default Keyboard