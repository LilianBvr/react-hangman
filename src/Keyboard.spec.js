import React from 'react'
import { mount, shallow } from 'enzyme'
import { expect } from 'chai'
import sinon from 'sinon'

import Keyboard from './Keyboard'
import { LETTERS } from './App'

describe(' <Keyboard /> ', () => {

    const onClickMock = sinon.spy()
    const props = {
        letters: LETTERS,
        testedLetters: [],
        onClick: () => {onClickMock()}
    }

    it('should render and match its reference snapshot', () => {
        expect(mount( <Keyboard {...props} />)).to.matchSnapshot()
    })

    describe('unit tests', () => {
        let wrapper

        beforeEach(() => {
            wrapper = shallow( <Keyboard {...props} />)
        })

        it('should call its onClick prop when clicking on a button', ()=>{

            wrapper.find('.letter').at(4).simulate('click')
            expect(onClickMock).to.have.been.called()
        })

        //TODO : verifier que les lettres test√©es sont en .hidden
    
    })

})