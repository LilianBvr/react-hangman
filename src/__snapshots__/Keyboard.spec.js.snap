// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` <Keyboard />  should render and match its reference snapshot 1`] = `
<Keyboard
  letters={
    Array [
      "A",
      "B",
      "C",
      "D",
      "E",
      "F",
      "G",
      "H",
      "I",
      "J",
      "K",
      "L",
      "M",
      "N",
      "O",
      "P",
      "Q",
      "R",
      "S",
      "T",
      "U",
      "V",
      "W",
      "X",
      "Y",
      "Z",
    ]
  }
  onClick={[Function]}
  testedLetters={Array []}
>
  <div
    className="keyboard"
  >
    <button
      className="letter "
      key="0"
      onClick={[Function]}
    >
      A
    </button>
    <button
      className="letter "
      key="1"
      onClick={[Function]}
    >
      B
    </button>
    <button
      className="letter "
      key="2"
      onClick={[Function]}
    >
      C
    </button>
    <button
      className="letter "
      key="3"
      onClick={[Function]}
    >
      D
    </button>
    <button
      className="letter "
      key="4"
      onClick={[Function]}
    >
      E
    </button>
    <button
      className="letter "
      key="5"
      onClick={[Function]}
    >
      F
    </button>
    <button
      className="letter "
      key="6"
      onClick={[Function]}
    >
      G
    </button>
    <button
      className="letter "
      key="7"
      onClick={[Function]}
    >
      H
    </button>
    <button
      className="letter "
      key="8"
      onClick={[Function]}
    >
      I
    </button>
    <button
      className="letter "
      key="9"
      onClick={[Function]}
    >
      J
    </button>
    <button
      className="letter "
      key="10"
      onClick={[Function]}
    >
      K
    </button>
    <button
      className="letter "
      key="11"
      onClick={[Function]}
    >
      L
    </button>
    <button
      className="letter "
      key="12"
      onClick={[Function]}
    >
      M
    </button>
    <button
      className="letter "
      key="13"
      onClick={[Function]}
    >
      N
    </button>
    <button
      className="letter "
      key="14"
      onClick={[Function]}
    >
      O
    </button>
    <button
      className="letter "
      key="15"
      onClick={[Function]}
    >
      P
    </button>
    <button
      className="letter "
      key="16"
      onClick={[Function]}
    >
      Q
    </button>
    <button
      className="letter "
      key="17"
      onClick={[Function]}
    >
      R
    </button>
    <button
      className="letter "
      key="18"
      onClick={[Function]}
    >
      S
    </button>
    <button
      className="letter "
      key="19"
      onClick={[Function]}
    >
      T
    </button>
    <button
      className="letter "
      key="20"
      onClick={[Function]}
    >
      U
    </button>
    <button
      className="letter "
      key="21"
      onClick={[Function]}
    >
      V
    </button>
    <button
      className="letter "
      key="22"
      onClick={[Function]}
    >
      W
    </button>
    <button
      className="letter "
      key="23"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="letter "
      key="24"
      onClick={[Function]}
    >
      Y
    </button>
    <button
      className="letter "
      key="25"
      onClick={[Function]}
    >
      Z
    </button>
  </div>
</Keyboard>
`;
